DOCSP-36996 - Authentication (#21)
mongodb/docs-pymongo: source/fundamentals/authentication.txt

.. _pymongo-auth:

Authentication Mechanisms
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: authorize, boto

MongoDB supports several different authentication mechanisms. This guide

> discusses all authentication methods currently supported by PyMongo.

Reviewer: ```suggestion
discusses all authentication methods currently supported by {+driver-short+}.
```

Author: i've started just leaving this as is, with plans to change on the second pass, since they don't affect the final page

.. code-block:: python

   >>> from pymongo import MongoClient
   >>> client = MongoClient('example.com',
   ...                      username='user',
   ...                      password='password',
   ...                      authSource='the_database',
   ...                      authMechanism='SCRAM-SHA-1')

Instead of passing these options directly to the ``MongoClient()`` constructor, you can
include them in your MongoDB URI:

.. code-block:: python

  >>> uri = "mongodb://user:password@example.com/?authSource=the_database&authMechanism=SCRAM-SHA-1"
  >>> client = MongoClient(uri)

.. tip::
  
   For best performance on Python versions older than 2.7.8, install the
   `backports.pbkdf2 <https://pypi.python.org/pypi/backports.pbkdf2/>`__ package.

MONGODB-CR
----------

.. warning::
  
  MONGODB-CR was deprecated with the release of MongoDB 3.6 and
  is no longer supported by MongoDB 4.0.

Before MongoDB 3.0, the default authentication mechanism was the "MongoDB
Challenge-Response" protocol (MONGODB-CR). As with the preceding authentication
mechanisms, you can pass the username, password, database, and authentication mechanism

> to the ``MongoClient()`` constructor or include them in your MongoDB URI:

Reviewer: ```suggestion
to the ``MongoClient()`` constructor, or include them in your MongoDB URI:
```

.. code-block:: python

   >>> from pymongo import MongoClient
   >>> client = MongoClient('example.com',
   ...                      username='user',
   ...                      password='password',
   ...                      authMechanism='MONGODB-CR')
   >>>
   >>> uri = "mongodb://user:password@example.com/?authSource=the_database&authMechanism=MONGODB-CR"
   >>> client = MongoClient(uri)

Default Authentication Mechanism
--------------------------------

> If you don't specify an authentication mechanism, PyMongo defaults to SCRAM-SHA-1 when

Reviewer: ```suggestion
If you don't specify an authentication mechanism, {+driver-short+} defaults to SCRAM-SHA-1 when
```

> connected to MongoDB 3.6. When connected to MongoDB 4.0+, PyMongo negotiates the mechanism

Reviewer: ```suggestion
connected to MongoDB 3.6. When connected to MongoDB 4.0+, {+driver-short+} negotiates the mechanism
```

.. code-block:: python

   >>> uri = "mongodb://user:password@example.com/default_db?authSource=admin"
   >>> client = MongoClient(uri)

> PyMongo will authenticate against the ``admin`` database, but the default database

Reviewer: ```suggestion
{+driver-short+} will authenticate against the ``admin`` database, but the default database
```

will be ``default_db``:

.. code-block:: python

   >>> # get_database with no "name" argument chooses the DB from the URI
   >>> db = MongoClient(uri).get_database()
   >>> print(db.name)
   'default_db'

.. _mongodb_x509:

MONGODB-X509
------------

The MONGODB-X509 mechanism authenticates via the X.509 certificate presented
by the driver during TLS/SSL negotiation. This authentication method requires
the use of TLS/SSL connections with certificate validation:

.. code-block:: python

   >>> from pymongo import MongoClient
   >>> client = MongoClient('example.com',
   ...                      authMechanism="MONGODB-X509",
   ...                      tls=True,
   ...                      tlsCertificateKeyFile='/path/to/client.pem',
   ...                      tlsCAFile='/path/to/ca.pem')

> MONGODB-X509 authenticates against the ``l$external`` virtual database, so you

Reviewer: ```suggestion
MONGODB-X509 authenticates against the ``$external`` virtual database, so you
```

Author: whoops!

do not have to specify a database in the URI:

.. code-block:: python

   >>> uri = "mongodb://example.com/?authMechanism=MONGODB-X509"
   >>> client = MongoClient(uri,
   ...                      tls=True,
   ...                      tlsCertificateKeyFile='/path/to/client.pem',
   ...                      tlsCAFile='/path/to/ca.pem')
   >>>

.. _MONGODB-AWS:

MONGODB-AWS
-----------

The MONGODB-AWS authentication mechanism is available in MongoDB 4.4+ and
requires extra {+driver-short+} dependencies. To use MONGODB-AWS authentication,

> install the ``aws`` extra when you install the ``pymongo`` package:

Reviewer: S: We changed "extra" to "option" in another page

```suggestion
install the ``aws`` option when you install the ``pymongo`` package:
```

you can use the provided credentials by giving the associated IAM User
`sts:AssumeRoleWithWebIdentity <https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html>`__
permission.

The driver uses an ``AssumeRoleWithWebIdentity`` call to retrieve temporary credentials
if the following criteria are met:

- You specify the MONGODB-AWS authentication mechanism.
- You don't provide a username and password.
- The ``AWS_WEB_IDENTITY_TOKEN_FILE``, ``AWS_ROLE_ARN``, and
  optional ``AWS_ROLE_SESSION_NAME`` environment variables are set.
- The application is using ``pymongo_auth_aws`` >= 1.1.0 for EKS support.

ECS Container
~~~~~~~~~~~~~

Applications can authenticate from an ECS container by using temporary
credentials assigned to the machine. The following example shows a sample URI on an ECS
container:

.. code-block:: python

   >>> from pymongo import MongoClient
   >>> uri = "mongodb+srv://example.mongodb.com/?authMechanism=MONGODB-AWS"
   >>> client = MongoClient(uri)

.. note::
  
   When using an ECS container for authentication, don't include the username, password,
   or session token in the URI.

>    PyMongo queries the ECS container endpoint to obtain these credentials.

Reviewer: ```suggestion
   {+driver-short+} queries the ECS container endpoint to obtain these credentials.
```

EC2 Instance
~~~~~~~~~~~~

Applications can authenticate from an EC2 instance by using temporary
credentials assigned to the machine. The following example shows a sample URI on an EC2
machine:

.. code-block:: python

   >>> from pymongo import MongoClient
   >>> uri = "mongodb+srv://example.mongodb.com/?authMechanism=MONGODB-AWS"
   >>> client = MongoClient(uri)

.. note::
  

>    When using an ECS instance for authentication, don't include the username, password,

Reviewer: ```suggestion
   When using an EC2 instance for authentication, don't include the username, password,
```

   or session token in the URI.

>    PyMongo queries the ECS container endpoint to obtain these credentials.

Reviewer: ```suggestion
   {+driver-short+} queries the ECS container endpoint to obtain these credentials.
```
